codec.go

```diff
@@ -49,6 +51,30 @@ type Codec interface {
 	Unmarshal([]byte, any) error
 }
 
+// StableCodec is an extension to Codec for serializing with stable output.
+type StableCodec interface {
+	// MarshalStable marshals the given message with stable field ordering.
+	//
+	// MarshalStable should return the same output for a given input. Although
+	// it is not guaranteed to be canonicalized, the marshalling routine for
+	// MarshalStable will opt for the most normalized output available for a
+	// given serialization.
+	//
+	// For practical reasons, it is possible for MarshalStable to return two
+	// different results for two inputs considered to be "equal" in their own
+	// domain, and it may change in the future with codec updates, but for
+	// any given concrete value and any given version, it should return the
+	// same output.
+	MarshalStable(any) ([]byte, error)
+
+	// IsBinary returns true if the marshalled data is binary for this codec.
+	//
+	// If this function returns false, the data returned from Marshal and
+	// MarshalStable are considered valid text and may be used in contexts
+	// where text is expected.
+	IsBinary() bool
```

Comment:
Codecs shouldn't need to expose anything about their format.

Reply:
Agreed with John here. To make Envoy and other middleboxes work, the runtime needs to know whether the encoded data is text or binary.